display_list/dl_shared.h

```diff
@@ -0,0 +1,508 @@
+                std::is_convertible<U*, T*>::value>::type>
+  dl_shared(dl_shared<U>&& that) : shareable_(that.release()) {}
+
+  ~dl_shared() {  //
+    capture(nullptr, false);
+  }
+
+  dl_shared<T>& operator=(std::nullptr_t) {  //
+    capture(nullptr, false);
+    return *this;
+  }
+
+  // Assign from raw pointer creates a new shared reference
+  dl_shared<T>& operator=(T* that) {
+    this->capture(that, true);
+    return *this;
+  }
+  template <typename U,
+            typename = typename std::enable_if<
+                std::is_convertible<U*, T*>::value>::type>
+  dl_shared<T>& operator=(U* that) {
+    this->capture(that, true);
+    return *this;
+  }
+
+  // Assign from lvalue creates a new shared reference
+  dl_shared<T>& operator=(const dl_shared<T>& that) {
+    this->capture(that.get(), true);
+    return *this;
+  }
+  template <typename U,
+            typename = typename std::enable_if<
+                std::is_convertible<U*, T*>::value>::type>
+  dl_shared<T>& operator=(const dl_shared<U>& that) {
+    this->capture(that.get(), true);
+    return *this;
+  }
+
+  // Assign from rvalue steals the reference
+  dl_shared<T>& operator=(dl_shared<T>&& that) {
+    this->capture(that.release(), false);
+    return *this;
+  }
+  template <typename U,
+            typename = typename std::enable_if<
+                std::is_convertible<U*, T*>::value>::type>
+  dl_shared<T>& operator=(dl_shared<U>&& that) {
+    this->capture(that.release(), false);
+    return *this;
+  }
+
+  T& operator*() const {
+    FML_DCHECK(shareable_ != nullptr);
+    return *shareable_;
+  }
+
+  explicit operator bool() const { return shareable_ != nullptr; }
+
+  T* get() const {
+    // NOLINTNEXTLINE(clang-analyzer-cplusplus.NewDelete)
+    return shareable_;
+  }
+  T* operator->() const { return shareable_; }
+  auto& operator[](int i) const { return (*shareable_)[i]; }
+
+ private:
+  T* shareable_;
+
+  [[nodiscard]] T* release() {
+    T* ret = shareable_;
+    shareable_ = nullptr;
+    return ret;
+  }
+  template <typename>
+  friend class dl_shared;
+
+  void capture(T* value, bool add_ref = false) {
+    if (add_ref && value) {
+      value->AddStrongRef();
+    }
+    T* old_shareable = shareable_;
+    shareable_ = value;
+    if (old_shareable) {
+      // NOLINTNEXTLINE(clang-analyzer-cplusplus.NewDelete)
+      old_shareable->RemoveStrongRef();
+    }
+  }
+};
+
+/// ----------------------------------------------------------------------
+/// A template class to manage weak shared pointers to objects which
+/// implement a compatible API, such as the DlShareable class defined
+/// below.
+///
+/// Assigning a pointer into a |dl_weak_shared| instance will automatically
+/// add a weak reference to the object which does not indicate an active
+/// interest in maintaining the object, but will nonetheless prevent its
+/// destruction until all of both weak and strong shared references are
+/// destroyed or replaced.
```

Comment:
I might be misreading this but this doesn't seem to be what a weak pointer is.

IOW, if I have this code:

```cpp
auto strong = std::make_shared<int>(1);
std::weak_ptr<int> weak= strong;
strong.reset();
```

I'd expect `weak` to be `nullptr` now. This description and implementation leads me to believe that's not the case.

Reply:
That's how the weak version of sk_sp is defined.

Reply:
dl_shared and sk_sp aren't full objects, they are templated wrappers around a pointer that ensure that proper ref-counting happens to the object.

Reply:
This could potentially be useful for caches.  The cache can hold a weak reference to the object and later scan the entries for objects that have no remaining strong references and therefore can be evicted.

I agree that the "weak reference" naming may mislead users into thinking that this behaves like `std::weak_ptr` where a weak reference will not prevent the object from being destructed.

Reply:
We already have caches like this using `std::weak_ptr` which have the extra advantage of not needing to wait for the scan operation to free memory in the cache (Although you still need to make sure to let go of the weak_ptr object itself), e.g. in https://github.com/flutter/engine/blob/293cab1a7dccdbbdf0303a00095b86b2f62b8850/common/graphics/texture.h#L106. [Moving to std::weak_ptr](https://github.com/flutter/engine/commit/e0625641f3fb677544d9585b6a32b46a58f962c1) resulted in a pretty marked performance benefit in the case of the volatile path tracker.

Reply:
There are different uses of these in caches. If it is the key in the cache, then you'd want it to automatically go away when the strong references are gone. If it is the value, then it might have a strong reference while that value is being used, but otherwise may live most of its useful life with only a weak reference from the cache waiting to be re-strengthened when you encounter its key again.

Reply:
While you're using the cache entry you call `.lock()` on the weak reference to find out if you can get a strong reference to it or not.

Reply:
It looks like Sk weak pointers are a sort of hybrid. On the surface, they work like I implemented `dl_weak_shared` in that the weak counts keep the object from being deleted. However, when the count goes to "all weak", they do call a virtual method that you could implement to dump your aggregate data so that while your base object remains alive, at least it doesn't hold on to any large resources when it is in the weak state. That is totally up to the object that inherits from the `SkWeakRefCnt` base class, though.

Reply:
> I might be misreading this but this doesn't seem to be what a weak pointer is.
> 
> IOW, if I have this code:
> 
> ```c++
> auto strong = std::make_shared<int>(1);
> std::weak_ptr<int> weak= strong;
> strong.reset();
> ```
> 
> I'd expect `weak` to be `nullptr` now. This description and implementation leads me to believe that's not the case.

The weak pointer there would behave as if it were null at that point, but the object would not have been freed. That is the default behavior of the `dl_weak_shared` as I prototyped it, but I'm not planning to go with that prototype anyway.

Basically, `make_shared` will allocate the std control block inline with the object. When the strong reference goes away, the control block must survive which means the control block cannot be freed which means the object memory cannot be freed either.

The `dl_weak_shared` that I originally wrote, just as an exercise in prototyping, also had a similar behavior in that `strengthen()` would return `nullptr` unless you passed an argument that indicated whether you wanted to strengthen it even if it was a zombie. The object wouldn't be freed until all references disappeared - similar to the memory allocation behavior of your example.

Potential differences are that I haven't looked into when the `std` mechanism would call the destructor method on the object, but since it tends to use the default deleter, it may not be until it is ready to free the memory which means that it will wait until the last strong or weak reference is dropped. `sk_sp` at least provides a virtual method you can override to "clean up" the object when it enters the zombie state even if the memory itself is not going away.

Reply:
> Potential differences are that I haven't looked into when the `std` mechanism would call the destructor method on the object

It looks like the object destructor will be called whether the shared pointer was constructed using the manual constructor or via make_shared. This means that the only memory that is held after all strong references go away is the base allocation for the object body itself.